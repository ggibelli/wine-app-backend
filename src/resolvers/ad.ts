import Ads from '../data-sources/ads';
import Users from '../data-sources/users';
import { AdInput, AdInputUpdate, Resolvers } from '../generated/graphql';
// import { IAd } from '../models/ad';

interface StringIndexSignatureInterface {
  [index: string]: any;
}

interface MongoDataSource {
  ads: Ads;
  users: Users;
  // wines: Wines;
}

type StringIndexed<T> = T & StringIndexSignatureInterface;

export const resolver: StringIndexed<Resolvers> = {
  Ad: {
    __resolveType: (ad) => ad.typeProduct,
    postedBy(ad, _, { dataSources }: { dataSources: MongoDataSource }) {
      console.log('ooo');
      const postedBy = dataSources.users.getUser(ad.postedBy);
      console.log('chiamato');
      return postedBy;
    },
    activeNegotiations(ad) {
      if (!ad.negotiations) {
        return 0;
      }
      return ad.negotiations?.length;
    },
    numberViews(ad) {
      if (!ad.viewedBy) {
        return 0;
      }
      return ad.viewedBy?.length;
    },
    // wine(ad: IAd, _: void, { dataSources }: { dataSources: MongoDataSource }) {
    //   return dataSources.wines.getWine(ad.wineName)
    // },
  },
  AdWine: {
    postedBy(ad, _, { dataSources }: { dataSources: MongoDataSource }) {
      console.log('ooo');
      const postedBy = dataSources.users.getUser(ad);

      return postedBy;
    },
  },

  // AdWine: {
  // wine(ad, { dataSources }: { dataSources: MongoDataSource }) {
  //   return dataSources.wines.getWine(ad.wineName)
  // },
  // },

  //AdGrape: {
  // vineyard(ad, { dataSources }: { dataSources: MongoDataSource }) {
  //   return dataSources.vineyard.getVineyard(ad.vineyardName)
  // },
  //},

  Query: {
    ads(_, args, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.ads.getAds(args);
    },
    ad(_, args, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.ads.getAd(args._id);
    },
  },

  Mutation: {
    createAd(
      _: any,
      { input }: { input: AdInput },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      const createdAd = dataSources.ads.createAd(input);
      return createdAd;
    },
    updateAd(
      _: any,
      { input }: { input: AdInputUpdate },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      const updateAd = dataSources.ads.updateAdWine(input);
      return updateAd;
    },
    deleteAd(
      _: any,
      { id }: { id: string },
      { dataSources }: { dataSources: MongoDataSource }
    ) {
      const removedAd = dataSources.ads.deleteAdWine(id);
      return removedAd;
    },
  },
};
