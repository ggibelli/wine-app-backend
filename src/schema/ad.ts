import { gql } from 'apollo-server-express';

export const typeDefs = gql`
  directive @date(defaultFormat: String = "mmmm d, yyyy") on FIELD_DEFINITION
  directive @authenticated on FIELD_DEFINITION
  directive @authorized on FIELD_DEFINITION
  type Address {
    via: String!
    CAP: Int!
    comune: String!
    provincia: Province!
    regione: Regioni!
  }

  input AddressInput {
    via: String!
    CAP: Int!
    comune: String!
    provincia: Province!
    regione: Regioni!
  }

  input AdInput {
    typeAd: TypeAd!
    typeProduct: TypeProduct!
    content: String!
    address: AddressInput!
    harvest: Int!
    abv: Float!
    wineName: String
    vineyardName: String
    sottoZona: String
    menzione: Menzione
    metodoProduttivo: MetodoProduttivo
    priceFrom: Float!
    priceTo: Float!
    litersFrom: Int
    litersTo: Int
    kgFrom: Int
    kgTo: Int
  }

  input AdInputUpdate {
    _id: ID!
    wineName: String
    vineyardName: String
    sottoZona: String
    menzione: Menzione
    metodoProduttivo: MetodoProduttivo
    harvest: Int
    abv: Float
    priceFrom: Float
    priceTo: Float
    litersFrom: Int
    litersTo: Int
    content: String
    address: AddressInput
  }

  interface Ad {
    _id: ID!
    postedBy: User! @authenticated
    harvest: Int!
    abv: Float!
    priceFrom: Float! @authenticated
    priceTo: Float! @authenticated
    content: String!
    address: Address! @authenticated
    "negotiations: [Negotiation]"
    activeNegotiations: Int @authenticated
    "viewedBy: [User]"
    numberViews: Int @authenticated
    typeAd: TypeAd!
    typeProduct: TypeProduct!
    isActive: Boolean!
    datePosted: Date! @date
  }

  type AdWine implements Ad {
    _id: ID!
    postedBy: User! @authenticated
    wineName: String!
    wine: Wine!
    sottoZona: String
    menzione: Menzione
    metodoProduttivo: MetodoProduttivo
    harvest: Int!
    abv: Float!
    priceFrom: Float!
    priceTo: Float!
    litersFrom: Int
    litersTo: Int
    content: String!
    address: Address!
    "negotiations: [Negotiation]"
    activeNegotiations: Int
    "viewedBy: [User]"
    numberViews: Int
    typeAd: TypeAd!
    typeProduct: TypeProduct!
    isActive: Boolean!
    datePosted: Date! @date
  }

  type AdGrape implements Ad {
    _id: ID!
    postedBy: User! @authenticated
    vineyardName: String!
    vineyard: Vineyard!
    harvest: Int!
    abv: Float!
    priceFrom: Float!
    priceTo: Float!
    kgFrom: Int!
    kgTo: Int!
    content: String!
    address: Address!
    "negotiations: [Negotiation]"
    activeNegotiations: Int
    "viewedBy: [User]"
    numberViews: Int
    typeAd: TypeAd!
    typeProduct: TypeProduct!
    isActive: Boolean!
    datePosted: Date! @date
  }

  type Query {
    ad(_id: ID!): Ad
    ads(
      typeAd: TypeAd!
      typeProduct: TypeProduct!
      wineName: String
      vineyardName: String
      skip: Int
      orderBy: QueryOrderBy
      limit: Int
    ): [Ad]
  }

  type Mutation {
    createAd(input: AdInput!): Ad @authenticated
    updateAd(input: AdInputUpdate!): Ad @authenticated
    deleteAd(id: ID!): Ad @authenticated
  }
  type Subscription {
    adPosted: Ad! @authenticated
  }
`;
