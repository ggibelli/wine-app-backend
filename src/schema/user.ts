import { gql } from 'apollo-server-express';

export const typeDefs = gql`
  type ProducedWines {
    wine: Wine!
    bottlesProduced: Int!
    metodoProduttivo: MetodoProduttivo!
  }

  type OwnedVineyards {
    vineyard: Vineyard!
    tonsProduced: Int!
    metodoProduttivo: MetodoProduttivo!
  }

  input ProducedWinesInput {
    wine: ID!
    bottlesProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  input OwnedVineyardsInput {
    vineyard: ID!
    tonsProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  input UserInput {
    email: String!
    password: String!
    firstName: String!
    lastName: String!
    pIva: String!
    phoneNumber: String!
    address: AddressInput!
    hideContact: Boolean!
    #verified: Boolean
    #premium: Boolean
    #ads: [ID!]
    #negotiations: [ID!]
    #reviews: [ID!]
    #adsRemaining: Int
    producedWines: ProducedWinesInput
    ownedVineyards: OwnedVineyardsInput
  }

  input UserInputUpdate {
    _id: ID!
    email: String
    password: String
    firstName: String
    lastName: String
    pIva: String
    phoneNumber: String
    address: AddressInput
    isVerified: Boolean
    isPremium: Boolean
    hideContact: Boolean!
    producedWines: ProducedWinesInput
    ownedVineyards: OwnedVineyardsInput
  }

  type Coordinates {
    latitude: Float
    longitude: Float
  }

  type User {
    _id: ID!
    email: String!
    firstName: String!
    lastName: String!
    pIva: String!
    phoneNumber: String!
    address: Address!
    isVerified: Boolean!
    coordinates: Coordinates
    isPremium: Boolean
    isAdmin: Boolean!
    hideContact: Boolean!
    ads: [Ad!]
    savedAds: [Ad!]
    messages: [Message!]
    negotiations: [Negotiation!]
    reviews: [Review!]
    adsRemaining: Int @authorized
    dateCreated: Date! @date
    producedWines: ProducedWines
    ownedVineyards: OwnedVineyards
  }

  type AuthUser {
    token: String!
    user: User!
  }

  type AuthUserPayload {
    response: AuthUser
    errors: [Errors]
  }

  type UserPayload {
    response: User
    errors: [Errors]
  }

  extend type Query {
    users: [User!] @authenticated
    user(id: ID!): User @authenticated
    me: User @authenticated
  }

  extend type Mutation {
    createUser(user: UserInput!): AuthUserPayload
    updateUser(user: UserInputUpdate!): UserPayload @authenticated
    deleteUser(id: ID!): UserPayload @authenticated
    login(email: String!, password: String!): AuthUserPayload
  }
`;
