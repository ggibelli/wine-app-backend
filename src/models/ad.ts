/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import mongoose, { Schema, Document } from 'mongoose';
// import { Address } from '../types';
import {
  // Province,
  // Regioni,
  METODOPRODUTTIVO,
} from '../utils/enumMongooseHelper';
import { IUserDoc } from './user';
import { IWineDoc } from './wine';
import { TypeAd, TypeProduct, Menzione } from '../types';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { Negotiation } from '../generated/graphql';

export interface IAd {
  postedBy: IUserDoc['_id'];
  wineName?: string;
  wine?: IWineDoc['_id'];
  sottoZona?: string;
  menzione?: Menzione;
  metodoProduttivo?: MetodoProduttivo;
  vineyardName?: string;
  vineyard?: IVineyardDoc['_id'];
  harvest: number;
  abv: number;
  priceFrom: number; //se vendita priceFrom e priceTo settati a stesso numero
  priceTo: number;
  litersFrom?: number; //se vendite litersFrom e litersTo settati a stesso numero
  litersTo?: number;
  kgFrom?: number;
  kgTo?: number;
  content?: string | null | undefined;
  // address: Address;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'annuncio, solo attive, solo graphql??
  viewedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  savedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  typeAd: TypeAd;
  typeProduct: TypeProduct;
  needsFollowUp: boolean;
  isActive: boolean;
  datePosted: Date;
}

export interface AdGraphQl extends IAd {
  _id: mongoose.Types.ObjectId | string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const adSchemaFields: Record<keyof IAd, any> = {
  postedBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  typeAd: {
    type: String,
    enum: [TypeAd.SELL, TypeAd.BUY],
    required: true,
  },
  typeProduct: {
    type: String,
    enum: [TypeProduct.ADWINE, TypeProduct.ADGRAPE],
    required: true,
  },
  wineName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
    index: true,
  },
  wine: {
    type: Schema.Types.ObjectId,
    ref: 'Wine',
    // required() {
    //   return this.typeProduct === TypeProduct.ADWINE;
    // },
  },
  sottoZona: {
    type: String,
  },
  menzione: {
    type: String,
    enum: ['CLASSICO', 'RISERVA', 'SUPERIORE', 'VIGNA'],
  },
  metodoProduttivo: {
    type: String,
    enum: Object.values(METODOPRODUTTIVO),
    required: true,
    default: 'CONVENZIONALE',
  },
  vineyardName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
    index: true,
  },
  vineyard: {
    type: Schema.Types.ObjectId,
    ref: 'Vineyard',
    // required() {
    //   return this.typeProduct === TypeProduct.ADGRAPE;
    // },
  },
  abv: Number,
  harvest: {
    type: Number,
    required: true,
  },
  priceFrom: {
    type: Number,
    required: true,
  },
  priceTo: {
    type: Number,
    required: true,
  },
  litersFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  litersTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  kgFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  kgTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  // address: {
  //   comune: {
  //     type: String,
  //     required: true,
  //   },
  //   provincia: {
  //     type: String,
  //     enum: Object.values(Province),
  //     required: true,
  //   },
  //   regione: {
  //     type: String,
  //     enum: Object.values(Regioni),
  //     required: true,
  //   },
  // },
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  viewedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  savedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  content: {
    type: String,
  },
  isActive: {
    type: Boolean,
    default: true,
  },
  needsFollowUp: {
    type: Boolean,
    default: true,
  },
  datePosted: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

export interface IAdDoc extends IAd, Document {}

const adSchema = new Schema(adSchemaFields);

adSchema.index({ typeProduct: 1, typeAd: 1, wineName: 1, vineyardName: 1 });

export const Ad = mongoose.model<IAdDoc>('Ad', adSchema);
