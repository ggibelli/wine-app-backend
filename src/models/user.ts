/* eslint-disable @typescript-eslint/no-unsafe-return */
import mongoose, { Schema, Document, HookNextFunction } from 'mongoose';
import mongooseUniqueValidator from 'mongoose-unique-validator';
import { Address } from '../types';
import { Province, Regioni } from '../utils/enumMongooseHelper';
import { IAdDoc } from './ad';
import { IWineDoc } from './wine';
import { IReviewDoc } from './review';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { METODOPRODUTTIVO } from '../utils/enumMongooseHelper';
import bcrypt from 'bcrypt';
import { IMessageDoc } from './message';
import { Negotiation } from '../generated/graphql';

const HASH_ROUNDS = 10;

interface ProducedWines {
  wine: IWineDoc['_id'] | IWineDoc;
  bottlesProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

interface OwnedVineyards {
  vineyard: IVineyardDoc['_id'] | IVineyardDoc;
  tonsProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface IUser {
  email: string;
  firstName: string;
  lastName: string;
  password: string;
  pIva: string;
  phoneNumber: string;
  address: Address;
  coordinates?: Coordinates;
  isVerified: boolean;
  isPremium?: boolean;
  isAdmin: boolean;
  hideContact: boolean;
  ads?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  savedAds?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  messages?: mongoose.Types.Array<IMessageDoc['_id'] | IMessageDoc>;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'utente, attive e non, concluse e non
  reviews?: mongoose.Types.Array<IReviewDoc['_id'] | IReviewDoc>; // recensioni fatte e ricevute dall'utente
  adsRemaining?: number;
  dateCreated: Date;
  producedWines?: mongoose.Types.Array<ProducedWines>;
  ownedVineyards?: mongoose.Types.Array<OwnedVineyards>;
}

export interface IUserDoc extends IUser, Document {
  validatePassword(password: string): Promise<boolean>;
}

export interface UserGraphQl extends IUser {
  validatePassword(password: string): Promise<boolean>;
  _id: mongoose.Types.ObjectId;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const userSchemaFields: Record<keyof IUser, any> = {
  email: {
    type: String,
    unique: true,
    minlength: 3,
    required: true,
  },
  phoneNumber: {
    type: String,
    unique: true,
    minlength: 10,
    required: true,
  },
  firstName: String,
  lastName: String,
  password: String,
  pIva: {
    type: String,
    unique: true,
    minlength: 11,
    required: true,
  },
  address: {
    via: {
      type: String,
      required: true,
      minlength: 5,
    },
    CAP: {
      type: String,
      required: true,
      minlength: 5,
    },
    comune: { type: String, required: true, minlength: 5 },
    provincia: {
      type: String,
      enum: Object.values(Province),
      required: true,
    },
    regione: {
      type: String,
      enum: Object.values(Regioni),
      required: true,
    },
  },
  coordinates: {
    latitude: Number,
    longitude: Number,
  },
  isVerified: {
    type: Boolean,
    required: true,
    default: false,
  },
  isPremium: {
    type: Boolean,
    default: false,
  },
  isAdmin: {
    type: Boolean,
    default: false,
  },
  hideContact: {
    type: Boolean,
    default: true,
  },
  ads: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  savedAds: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  reviews: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Review',
    },
  ],
  adsRemaining: Number,
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  producedWines: [
    {
      wine: {
        type: Schema.Types.ObjectId,
        ref: 'Wine',
      },
      bottlesProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
  ownedVineyards: [
    {
      vineyard: {
        type: Schema.Types.ObjectId,
        ref: 'Vineyard',
      },
      tonsProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
};

const userSchema = new Schema(userSchemaFields);

userSchema.pre('save', async function (next: HookNextFunction) {
  // here we need to retype 'this' because by default it is
  // of type Document from which the 'IUser' interface is inheriting
  // but the Document does not know about our password property
  const thisObj = this as IUserDoc;

  if (!this.isModified('password')) {
    return next();
  }

  try {
    const salt = await bcrypt.genSalt(HASH_ROUNDS);
    thisObj.password = await bcrypt.hash(thisObj.password, salt);
    return next();
  } catch (e) {
    return next(e);
  }
});

userSchema.methods.validatePassword = async function (pass: string) {
  return bcrypt.compare(pass, this.password);
};

userSchema.plugin(mongooseUniqueValidator);

export const User = mongoose.model<IUserDoc>('User', userSchema);
