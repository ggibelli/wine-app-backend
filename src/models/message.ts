import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { INegotiationDoc } from './negotiation';

export interface IMessage {
  negotiation: INegotiationDoc['_id'];
  content: string;
  sentBy: IUserDoc['_id'];
  sentTo: IUserDoc['_id'];
  dateSent: Date;
  isViewed: boolean;
}

export interface MessageGraphQl extends IMessage {
  _id: mongoose.Types.ObjectId;
}

export interface IMessageDoc extends IMessage, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const messageSchemaFields: Record<keyof IMessage, any> = {
  negotiation: {
    type: Schema.Types.ObjectId,
    ref: 'Negotiation',
    required: true,
  },
  content: {
    type: String,
    minlength: 1,
    required: true,
  },
  sentBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  sentTo: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  dateSent: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  isViewed: {
    type: Boolean,
    default: false,
  },
};

const messageSchema = new Schema(messageSchemaFields);

messageSchema.index({ negotiation: 1, sentTo: 1, sentFrom: 1 });

export const Message = mongoose.model<IMessageDoc>('Message', messageSchema);
