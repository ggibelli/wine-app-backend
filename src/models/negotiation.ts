/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { IAdDoc } from './ad';
import { IMessageDoc } from './message';
import { TypeAd } from '../types';

export interface INegotiation {
  createdBy: IUserDoc['_id'];
  ad: IAdDoc['_id'];
  forUserAd: IUserDoc['_id'] | IUserDoc;
  messages?: Array<IMessageDoc['_id'] | IMessageDoc>;
  isConcluded: boolean;
  dateCreated: Date;
  type: TypeAd;
}

export interface INegotiationDoc extends INegotiation, Document {}

const negotiationSchemaFields: Record<keyof INegotiation, any> = {
  createdBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  type: {
    type: String,
    enum: ['Sell', 'Buy'],
    required: true,
  },
  ad: {
    type: Schema.Types.ObjectId,
    ref: 'Ad',
    required: true,
  },
  forUserAd: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  isConcluded: {
    type: Boolean,
    default: false,
  },
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

const negotiationSchema = new Schema(negotiationSchemaFields);

negotiationSchema.index({ createdBy: 1, ad: 1 }, { unique: true });

export const Negotiation = mongoose.model<INegotiationDoc>(
  'Negotiation',
  negotiationSchema
);
