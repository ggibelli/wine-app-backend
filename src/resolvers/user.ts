import {
  AdsResult,
  NegotiationResult,
  Resolvers,
  ReviewResult,
  UserAdsArgs,
  UserInput,
  UserInputUpdate,
  UserNegotiationsArgs,
  UserReviewsArgs,
} from '../generated/graphql';
import { UserGraphQl } from '../models/user';
import Users, { AuthResponse, UserResponse } from '../data-sources/users';
import Ads from '../data-sources/ads';
import Messages from '../data-sources/messages';
import Negotiations from '../data-sources/negotiations';
import { ForbiddenError } from 'apollo-server-express';
import Reviews from '../data-sources/reviews';
import { IAdDoc } from '../models/ad';
import { MessageGraphQl } from '../models/message';

interface StringIndexSignatureInterface {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [index: string]: any;
}

type StringIndexed<T> = T & StringIndexSignatureInterface;

interface MongoDataSource {
  users: Users;
  ads: Ads;
  messages: Messages;
  negotiations: Negotiations;
  reviews: Reviews;
  // wines: Wines;
}

export const resolver: StringIndexed<Resolvers> = {
  Query: {
    async users(
      _,
      __,
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<UserGraphQl[]> {
      return dataSources.users.getUsers();
    },
    async user(
      _,
      { id }: { id: string },
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<UserGraphQl | undefined | null> {
      return dataSources.users.getUser(id);
    },
    me(_, __, { user }: { user: UserGraphQl }): UserGraphQl {
      console.log('yo');
      return user;
    },
  },
  Mutation: {
    async createUser(
      _,
      { user }: { user: UserInput },
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<AuthResponse> {
      return dataSources.users.createUser(user);
    },
    async updateUser(
      _,
      { user }: { user: UserInputUpdate },
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<UserResponse> {
      return dataSources.users.updateUser(user);
    },
    async deleteUser(
      _,
      { id }: { id: string },
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<UserResponse> {
      return dataSources.users.deleteUser(id);
    },
    async login(
      _,
      { email, password }: { email: string; password: string },
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<AuthResponse> {
      return dataSources.users.login(email, password);
    },
  },
  User: {
    async ads(
      user: UserGraphQl,
      args: UserAdsArgs,
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<AdsResult> {
      return dataSources.ads.getAdsByUser(user._id, args);
    },
    async messages(
      root: UserGraphQl,
      _,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ): Promise<MessageGraphQl[]> {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.messages.getMessages();
      }
      throw new ForbiddenError('You can only see your own messages');
    },
    async negotiations(
      root: UserGraphQl,
      args: UserNegotiationsArgs,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ): Promise<NegotiationResult> {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.negotiations.getNegotiations(args);
      }
      throw new ForbiddenError('You can only see your own negotiations');
    },
    async savedAds(
      root: UserGraphQl,
      _,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ): Promise<IAdDoc[]> {
      if (root._id.toString() === user._id.toString()) {
        if (user.savedAds?.length) {
          return dataSources.ads.findManyByIds(user.savedAds) as Promise<
            IAdDoc[]
          >;
        }
        return [];
      }
      throw new ForbiddenError('You can only see your own saved ads');
    },
    async reviews(
      root: UserGraphQl,
      args: UserReviewsArgs,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ): Promise<ReviewResult> {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.reviews.getReviews(args);
      }
      throw new ForbiddenError('You can only see your own negotiations');
    },
    email(root: UserGraphQl, _, { user }: { user: UserGraphQl }): string {
      if (root._id.toString() === user._id.toString()) {
        return root.email;
      }
      if (!root.hideContact) {
        return root.email;
      }
      return 'The user prefers to hide the email';
    },
    phoneNumber(root: UserGraphQl, _, { user }: { user: UserGraphQl }): string {
      if (root._id.toString() === user._id.toString()) {
        return root.phoneNumber;
      }
      if (!root.hideContact) {
        return root.phoneNumber;
      }
      return 'The user prefers to hide the phone number';
    },
    // aggiungo logica per mostrare solo negotiation e certi dati solo se sono i miei
  },
};
