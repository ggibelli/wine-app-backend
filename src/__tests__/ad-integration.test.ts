import { gql } from 'apollo-server-express';
import {
  testClient,
  connectToDb,
  dropTestDb,
  closeDbConnection,
} from '../tests/integrationSetup';
//import { ObjectId } from 'mongodb';
import { User } from '../models/user';
import { Ad } from '../models/ad';
import { users, ads } from '../tests/mocksTests';

const { query } = testClient;

const AD = gql`
  fragment AdPreview on Ad {
    #datePosted
    postedBy {
      firstName
    }
    harvest
    abv
    priceFrom
    priceTo
    ... on AdWine {
      wineName
      litersFrom
      litersTo
    }
    ... on AdGrape {
      vineyardName
      kgFrom
      kgTo
    }
  }
  {
    ads(typeAd: Buy, typeProduct: AdWine) {
      ...AdPreview
    }
  }
`;

beforeAll(async () => {
  await connectToDb();
  await dropTestDb();
  const usersMock = users();
  const adsMock = ads();
  const user = new User(usersMock[0]);
  const ad = new Ad({ ...adsMock[0], postedBy: user });
  await user.save();
  await ad.save();
});

describe('Integration test ads', () => {
  it.only('query ads wine in sale', async () => {
    const res = await query({ query: AD });
    expect(res).toMatchSnapshot();
  });
});

afterAll(async () => {
  await dropTestDb();
  await closeDbConnection();
});
