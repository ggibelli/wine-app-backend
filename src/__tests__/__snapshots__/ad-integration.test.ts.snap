// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test ads query ads wine in sale 1`] = `
Object {
  "data": Object {
    "ads": Array [
      null,
    ],
  },
  "errors": Array [
    [GraphQLError: You need to login to continue],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
